lib/vm/src/instance/ref.rs

```diff
@@ -109,114 +52,102 @@ impl InstanceRef {
-            + self.instance_layout.size_of_val(tracker)
+        mem::size_of_val(self) + self.instance_layout.size_of_val(tracker)
             - mem::size_of_val(&self.instance_layout)
             + self.as_ref().size_of_val(tracker)
             - mem::size_of_val(&self.instance)
     }
 }
+
+/// An `InstanceRef` is responsible to properly deallocate,
+/// and to give access to an `Instance`, in such a way that `Instance`
+/// is unique, can be shared, safely, across threads, without
+/// duplicating the pointer in multiple locations. `InstanceRef`
+/// must be the only “owner” of an `Instance`.
+///
+/// Consequently, one must not share `Instance` but
+/// `InstanceRef`. It acts like an Atomically Reference Counter
+/// to `Instance`. In short, `InstanceRef` is roughly a
+/// simplified version of `std::sync::Arc`.
+///
+/// This `InstanceRef` must be freed with [`InstanceRef::deallocate_instance`]
+/// if and only if it has been set correctly. The `Drop` implementation of
+/// [`InstanceRef`] calls its `deallocate_instance` method without
+/// checking if this  property holds, only when `Self.strong` is equal to 1.
+///
+/// Note for the curious reader: [`InstanceAllocator::new`]
+/// and [`InstanceHandle::new`] will respectively allocate a proper
+/// `Instance` and will fill it correctly.
+///
+/// A little bit of background: The initial goal was to be able to
+/// share an [`Instance`] between an [`InstanceHandle`] and the module
+/// exports, so that one can drop a [`InstanceHandle`] but still being
+/// able to use the exports properly.
+#[derive(Debug, PartialEq, Clone, MemoryUsage)]
+pub struct InstanceRef(Arc<InstanceInner>);
+
+impl InstanceRef {
+    /// Create a new `InstanceRef`. It allocates nothing. It fills
+    /// nothing. The `Instance` must be already valid and
+    /// filled.
+    ///
+    /// # Safety
+    ///
+    /// `instance` must a non-null, non-dangling, properly aligned,
+    /// and correctly initialized pointer to `Instance`. See
+    /// [`InstanceAllocator`] for an example of how to correctly use
+    /// this API.
+    pub(super) unsafe fn new(instance: NonNull<Instance>, instance_layout: Layout) -> Self {
+        Self(Arc::new(InstanceInner {
+            instance_layout,
+            instance,
+        }))
+    }
+
+    /// Get a reference to the `Instance`.
+    #[inline]
+    pub(crate) fn as_ref(&self) -> &Instance {
+        (&*self.0).as_ref()
+    }
+
+    #[inline]
+    /// Only succeeds if ref count is 1.
+    pub(super) fn as_mut(&mut self) -> Option<&mut Instance> {
+        Some(Arc::get_mut(&mut self.0)?.as_mut())
+    }
+
+    #[inline]
+    /// # Safety
+    /// It is the caller's responsibility to ensure exclusivity and synchronization of the
+    /// instance.
+    pub(super) unsafe fn as_mut_unchecked(&mut self) -> &mut Instance {
+        let ptr: *mut InstanceInner = Arc::as_ptr(&self.0) as *mut _;
+        (&mut *ptr).as_mut()
```

Comment:
Two possibilities here.

1️⃣  The first one is the following:

```suggestion
        ptr.as_mut().unwrap().as_mut()
```

not tested, but I was thinking about [`ptr::as_mut`](https://doc.rust-lang.org/std/primitive.pointer.html#method.as_mut)

Reply:
We can't use `as_mut` because `as_mut` can fail. This function can never fail, I'll look into the other way though

Reply:
Then possibility number 2 seems appropriate: https://github.com/wasmerio/wasmer/pull/2326#discussion_r636045237

Reply:
For what it's worth, the unstable method https://doc.rust-lang.org/std/sync/struct.Arc.html#method.get_mut_unchecked (what we really want) is very similar internally to what the code is currently doing, so I think it's correct as-is.

Reply:
Sorry, it's `Arc::get_mut`, https://doc.rust-lang.org/std/sync/struct.Arc.html#method.get_mut